aten/src/ATen/MetaInit.cpp

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// All rights reserved.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include <ATen/MetaInit.h>
+
+#include <cstdint>
+
+#include <ATen/Tensor.h>
+#include <ATen/core/dispatch/Dispatcher.h>
+#include <c10/core/DispatchKeySet.h>
+#include <torch/library.h>
+
+namespace at {
+namespace {
+
+constexpr DispatchKeySet after_meta_init_keyset =
+    DispatchKeySet{DispatchKeySet::FULL_AFTER, DispatchKey::MetaInit};
+
+void metaInitFallback(const OperatorHandle& op, DispatchKeySet ks, torch::jit::Stack* s) {
+  DisableMetaInitGuard guard{};
+
+  // Just redispatch, effectively noop.
+  op.redispatchBoxed(ks & after_meta_init_keyset, s);
+}
+
+// Used to support nested calls.
+thread_local std::size_t meta_init_level = 0;
+
+} // namespace
+
+void enableMetaInit(bool value) {
+  if (value) {
+    meta_init_level++;
```

Comment:
Is there a reason to support nested meta_init ctx managers? I'm wondering if it might just be simpler to error out in such cases?

Reply:
I wanted to make sure that the API is flexible for different use cases. If a framework such as PyTorch Lightning uses it internally while the user's code also have a use for it, not allowing nested calls might cause a failure beyond your control.